### PR TITLE
Add autoremove support to yum module

### LINKED ISSUE #18904: Support yum autoremove
_From @font on July 28, 2016 15:50_

##### ISSUE TYPE

- Feature Idea
##### COMPONENT NAME

- yum
##### ANSIBLE VERSION

```
ansible 2.1.0.0
```
##### OS / ENVIRONMENT

- Managing Linux
##### SUMMARY

At times it is necessary, particularly for cleanup plays, to remove all unneeded packages that were originally installed as dependencies. Currently this can be done via a `command: yum -y autoremove`. However, it would be best to incorporate it into the yum module, especially because ansible issues a `[WARNING]: Consider using yum module rather than running yum` when not using the yum module.

_Copied from original issue: ansible/ansible-modules-core#4273_

### PR DESCRIPTION
**Pull request by nrwahl2**
##### SUMMARY
Fixes #18904: Add autoremove support to yum module

Modeled partly after DNF, which already has autoremove functionality. I don't have access to a system with yum at the moment. Will test personally when I'm able.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/yum

##### ANSIBLE VERSION
```
ansible 2.4.0 (issue/18904 4dde3d7254) last updated 2017/08/26 22:52:35 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/reid/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/reid/git/ansible/lib/ansible
  executable location = /home/reid/git/ansible/bin/ansible
  python version = 2.7.13 (default, Jun 26 2017, 10:20:05) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
```

##### ADDITIONAL INFORMATION
N/A

### CODE REVIEW

**Comment by dagwieers** [comment]
@nrwahl2 I noticed you add the issue # in the title of your PR, but that's very confusing (and is not necessary if you simply want to reference/auto-close the issue). I think it's better to simply summarize what the PR does, and leave it to the description to what it fixes or why it's needed.

One of the reasons this is confusing is that if people use the title to list PR's, your titles might mess up the links.

I see that other people are now copying your behavior, sigh...

**Comment by nrwahl2** [comment]
@dagwieers I stopped doing that a day or two ago when I noticed that no one else here was doing it that way ;)

I had been doing them like this after reading https://help.github.com/articles/closing-issues-using-keywords/ and wanting to make sure the PRs were linked to the issues.

Man I hate that people are copying me as though I know what I'm doing. I've been here like 10 days haha.

**Comment by dagwieers** [comment]
@nrwahl2 No problem :) I'll notify the others, and I am going to rewrite the existing titles too if you don't mind ;-)

**Comment by nrwahl2** [comment]
Go for it.

Sent from ProtonMail mobile

-------- Original Message --------

File: lib/ansible/modules/packaging/os/yum.py [hunk 1]
```diff
@@ -33,78 +33,87 @@
         several packages this can accept a comma separated list of packages
         or (as of 2.0) a list of packages.
     required: true
-    aliases: [ pkg ]
+    aliases: ['pkg']
   exclude:
     description:
-      - "Package name(s) to exclude when state=present, or latest"
```
Comment (maxamillion) [inline lib/ansible/modules/packaging/os/yum.py:39]: There's a lot of churn on changing double to single quotes, I could understand wanting to be consistent but it seems outside the scope of this change.

File: lib/ansible/modules/packaging/os/yum.py [hunk 2]
```diff
@@ -33,78 +33,87 @@
...
-        use can also list the following: C(installed), C(updates), C(available) and C(repos)."
+      - 'Package name to run the equivalent of yum list <package> against. In addition to listing packages,
+        use can also list the following: C(installed), C(updates), C(available) and C(repos).'
   state:
     description:
       - Whether to install (C(present) or C(installed), C(latest)), or remove (C(absent) or C(removed)) a package.
-    choices: [ absent, installed, latest, present, removed ]
+    choices: ['absent', 'installed', 'latest', 'present', 'removed']
     default: present
+  autoremove:
```
Comment (maxamillion) [inline lib/ansible/modules/packaging/os/yum.py:50]: This should either have a note about the version of yum required to make this work (the version of yum in RHEL/CentOS6 does not have this functionality) or implement it in a way that makes use of `clean_requirements_on_remove` instead of the `autoremove` yum sub-command.

If you do go the route of note about version requirements, I think it would be a good idea to have a check in the code that will do something like `module.fail_json(msg="YUM Version {} Required for autoremove".format(YUM_MIN_VERSION))` if an older version is found.

**Review by maxamillion — CHANGES_REQUESTED** [review]

**Comment by maxamillion** [comment]
@nrwahl2 any update here?

**Comment by fessmage** [comment]
I think this PR can be closed, because autoremove funtionality was added recently in #43621
